components/shared/net/clientstorage/actors_child.rs

```diff
@@ -0,0 +1,277 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::cell::{Cell, RefCell};
+use std::rc::Rc;
+
+use ipc_channel::ipc::IpcSender;
+use log::debug;
+
+use super::actor_id::ClientStorageActorId;
+use super::child::ClientStorageChild;
+use super::mixed_msg::ClientStorageMixedMsg;
+use super::proxy::ClientStorageProxy;
+use super::routed_msg::ClientStorageRoutedMsg;
+use super::test_cursor_msg::ClientStorageTestCursorMsg;
+use super::test_msg::ClientStorageTestMsg;
+
+struct BoundState {
+    proxy: Rc<ClientStorageProxy>,
+    ipc_sender: IpcSender<ClientStorageRoutedMsg>,
+    actor_id: ClientStorageActorId,
+}
+
+impl BoundState {
+    pub fn send_mixed_message(&self, msg: ClientStorageMixedMsg) {
+        self.send_routed_message(ClientStorageRoutedMsg {
+            actor_id: self.actor_id,
+            data: msg,
+        });
+    }
+
+    pub fn send_routed_message(&self, msg: ClientStorageRoutedMsg) {
+        self.ipc_sender.send(msg).unwrap();
+    }
+}
+
+struct TopBoundState {
+    base: BoundState,
+    next_local_id: Cell<i64>,
+}
+
+impl TopBoundState {
+    pub fn new(
+        proxy: Rc<ClientStorageProxy>,
+        ipc_sender: IpcSender<ClientStorageRoutedMsg>,
+        actor_id: ClientStorageActorId,
+    ) -> Self {
+        Self {
+            base: BoundState {
+                proxy,
+                ipc_sender,
+                actor_id,
+            },
+            next_local_id: Cell::new(-1),
+        }
+    }
+
+    fn proxy(&self) -> &Rc<ClientStorageProxy> {
+        &self.base.proxy
+    }
+
+    fn ipc_sender(&self) -> &IpcSender<ClientStorageRoutedMsg> {
+        &self.base.ipc_sender
+    }
+
+    fn actor_id(&self) -> &ClientStorageActorId {
+        &self.base.actor_id
+    }
+
+    fn send_mixed_message(&self, msg: ClientStorageMixedMsg) {
+        self.base.send_mixed_message(msg);
+    }
+
+    fn next_local_id(&self) -> i64 {
+        let next_local_id = self.next_local_id.get();
+        self.next_local_id.set(next_local_id - 1);
+
+        next_local_id
+    }
+}
+
+type PongCallback = Box<dyn Fn()>;
```

Comment:
This is quite interesting because it reminds me of https://github.com/servo/servo/pull/37983.

You are using `Trusted` so it's safe, but I think `DomRoot` would have compiled as well and not be safe.